File: python_rpn_solver/solver/parse/parser.py
from ..token.token import Token
from ..token.literal_token import LiteralToken
from ..token.operator_token import OperatorToken, Operator, Associativity


def rpn_from_tokens(tokens: list[Token]) -> list[Token]:
    operator_stack: list[OperatorToken] = []
    rpn_solve_list: list[Token] = []

    for token in tokens:
        if isinstance(token, LiteralToken):
            rpn_solve_list.append(token)
        elif isinstance(token, OperatorToken):
            if token.operator == Operator.RPAREN:
                while True:
                    if len(operator_stack) == 0:
                        raise ValueError(f"Mismatched Parentheses in token list\n\nSolve List:{rpn_solve_list}\n\nOperator Stack:{operator_stack}")
                    if (last_operator := operator_stack.pop()).operator != Operator.LPAREN:
                        rpn_solve_list.append(last_operator)
                    else:
                        break
                    if len(operator_stack) == 0:
                        break
            else:
                while len(operator_stack) != 0 and precedence_compare(token, operator_stack[-1]):
                    rpn_solve_list.append(operator_stack.pop())
                operator_stack.append(token)
    while len(operator_stack) != 0:
        rpn_solve_list.append(operator_stack.pop())

    return rpn_solve_list


def solve_rpn(tokens: list[Token]) -> tuple[float, list[str]]:
    tokens = tokens.copy() # to not mess with the original parameters
    steps = ["Showing steps for RPN Solve:"]
    step = 0
    step_changed = True

    i = 0
    while i < len(tokens):
        token = tokens[i]
        if isinstance(token, OperatorToken):
            if i < token.numOperands:
                raise ValueError(f'rpn solve failed: Operator at index [{i}] too far left: missing arguments')
            else:
                if token.numOperands == 2:
                    lhs, rhs = tokens[i-2], tokens[i-1]
                    if not isinstance(lhs, LiteralToken) or not isinstance(rhs, LiteralToken):
                        raise ValueError('Operator arguments were not both literals')
                    arithmetic_result = 0
                    match token.operator:
                        case Operator.PLUS:
                            arithmetic_result = lhs.value + rhs.value
                        case Operator.MINUS:
                            arithmetic_result = lhs.value - rhs.value
                        case Operator.MULTIPLY:
                            arithmetic_result = lhs.value * rhs.value
                        case Operator.DIVIDE:
                            arithmetic_result = lhs.value / rhs.value
                        case Operator.EXPONENT:
                            arithmetic_result = pow(lhs.value, rhs.value)
                    step += 1
                    step_changed = True
                    steps.append(f"BINARY OPERATION {token.to_simple_string()} ON {lhs.to_simple_string()} AND {rhs.to_simple_string()} => INSERT {arithmetic_result}")
                    
                    tokens.remove(tokens[i])
                    tokens.remove(tokens[i-1])
                    tokens[i-2] = LiteralToken(arithmetic_result)
                    i -= 2
                elif token.numOperands == 1:
                    lhs = tokens[i-1]
                    if not isinstance(lhs, LiteralToken):
                        raise ValueError(f'Operator argument for op: {lhs.to_simple_string()} was not a literal')
                    arithmetic_result = 0
                    match token.operator:
                        case Operator.UNARY_PLUS:
                            arithmetic_result = +lhs.value
                        case Operator.UNARY_MINUS:
                            arithmetic_result = -lhs.value
                    step += 1
                    step_changed = True
                    steps.append(f'UNARY OPERATION {token.to_simple_string()} ON {lhs.to_simple_string()} => INSERT {arithmetic_result}')

                    tokens.remove(tokens[i])
                    tokens[i-1] = LiteralToken(arithmetic_result)
                    i -= 1
        if step_changed:
            steps.append(f"step: {step} list: {' '.join(token.to_simple_string() for token in tokens)}")
            step_changed = False
        i += 1
    if len(tokens) == 1 and isinstance(tokens[0], LiteralToken):
        return tokens[0].value, steps
    else:
        raise ValueError(f'rpn solve list did not evaluate to a single literal result, tokens left: {tokens}')


def precedence_compare(opToken: OperatorToken, otherOpToken: OperatorToken) -> bool:
    if otherOpToken.associativity == Associativity.LEFT:
        return opToken.precedence() <= otherOpToken.precedence()
    elif otherOpToken.associativity == Associativity.RIGHT:
        return opToken.precedence() < otherOpToken.precedence()
    else:
        return False # must be a paren

File: dartRpnSolver/parse/Parser.dart
import 'dart:collection';
import 'dart:math' as math;

import '../exceptions/AggregateException.dart';
import '../token/InvalidToken.dart';
import '../token/LiteralToken.dart';
import '../token/OperatorToken.dart';
import '../token/Token.dart';
import 'Tokenizer.dart';

double parse(String input) {
  final List<Token> tokens = tokenize(input);
  print(tokens.join(', '));
  final List<Token> rpnSolveList = _tokensToRpn(tokens);
  final SolveResult solveResult = _shuntingYardSolve(rpnSolveList);
  return solveResult.value;
}

List<Token> _tokensToRpn(List<Token> tokens) {
  tokens.throwIfContainsInvalid();

  final operatorStack = Queue<OperatorToken>();
  final rpnSolveList = <Token>[];

  for (final token in tokens) {
    if (token is LiteralToken) {
      rpnSolveList.add(token);
    } else if (token is OperatorToken) {
      if (token.operator == Operator.RPAREN) {
        while (true) {
          if (operatorStack.isEmpty) {
            throw ArgumentError(
                'Mismatched Parentheses in token list\n\nSolve List:$rpnSolveList\n\nOperator Stack:$operatorStack');
          }
          final poppedOperator = operatorStack.removeLast();
          if (poppedOperator.operator == Operator.LPAREN) {
            break;
          } else {
            rpnSolveList.add(poppedOperator);
          }
          if (operatorStack.isEmpty) {
            break;
          }
        }
      } else {
        while (operatorStack.isNotEmpty &&
            _precedenceCompare(token, operatorStack.last)) {
          rpnSolveList.add(operatorStack.removeLast());
        }
        operatorStack.addLast(token);
      }
    }
  }

  while (operatorStack.isNotEmpty) {
    rpnSolveList.add(operatorStack.removeLast());
  }

  return rpnSolveList;
}

bool _precedenceCompare(OperatorToken leftOpToken, OperatorToken rightOpToken) {
  switch (rightOpToken.associativity) {
    case Associativity.LEFT:
      return leftOpToken.precedence <= rightOpToken.precedence;
    case Associativity.RIGHT:
      return leftOpToken.precedence < rightOpToken.precedence;
    case Associativity.NONE:
      return false;
  }
}

class SolveResult {
  final double value;

  SolveResult(this.value);
}

SolveResult _shuntingYardSolve(List<Token> rpnSolveList) {
  rpnSolveList.throwIfContainsInvalid();

  final solveStack = Queue<Token>();

  for (final token in rpnSolveList) {
    if (token is LiteralToken) {
      solveStack.addLast(token);
    } else if (token is OperatorToken) {
      switch (token.numOperands) {
        case 2:
          final rhs = solveStack.removeLast();
          final lhs = solveStack.removeLast();

          if (lhs is! LiteralToken || rhs is! LiteralToken) {
            throw ArgumentError(
                'Operator arguments were not both literals. lhs: ${lhs} rhs: ${rhs}.');
          }

          final double arithmeticResult;
          switch (token.operator) {
            case Operator.PLUS:
              arithmeticResult = lhs.value + rhs.value;
              break;
            case Operator.MINUS:
              arithmeticResult = lhs.value - rhs.value;
              break;
            case Operator.MULTIPLY:
              arithmeticResult = lhs.value * rhs.value;
              break;
            case Operator.DIVIDE:
              arithmeticResult = lhs.value / rhs.value;
              break;
            case Operator.EXPONENT:
              arithmeticResult = math.pow(lhs.value, rhs.value).toDouble();
              break;
            default:
              throw ArgumentError(
                  'Invalid operator provided: ${token.operator}, number of operands: ${token.numOperands}. Binary operator expected.');
          }

          solveStack.addLast(LiteralToken(arithmeticResult));
          break;

        case 1:
          final operand = solveStack.removeLast();

          if (operand is! LiteralToken) {
            throw ArgumentError(
                'Operator argument was not a literal. operand: ${operand}.');
          }

          final double arithmeticResult;
          switch (token.operator) {
            case Operator.UNARY_PLUS:
              arithmeticResult = operand.value;
              break;
            case Operator.UNARY_MINUS:
              arithmeticResult = -operand.value;
              break;
            default:
              throw ArgumentError(
                  'Invalid operator provided: ${token.operator}, number of operands: ${token.numOperands}. Unary operator expected.');
          }

          solveStack.addLast(LiteralToken(arithmeticResult));
          break;

        default:
          throw ArgumentError(
              'Invalid number of operands ${token.numOperands} for operator ${token.operator}');
      }
    }
  }

  if (solveStack.length != 1) {
    throw ArgumentError(
        'Final result was not a single value. Result stack count: ${solveStack.length}, Result stack values: ${solveStack.map((t) => t).join(',')}');
  }

  final finalToken = solveStack.single;
  if (finalToken is! LiteralToken) {
    throw StateError('Final token is not a LiteralToken');
  }

  return SolveResult(finalToken.value);
}

extension TokenListValidation on List<Token> {
  void throwIfContainsInvalid() {
    final invalidTokens = whereType<InvalidToken>().toList();
    if (invalidTokens.isNotEmpty) {
      final errors = invalidTokens
          .map((token) => ArgumentError(
              'Invalid token: ${token.invalidShard} at position ${token.position}'))
          .toList();
      throw AggregateException(
          'Parsing errors encountered:', errors.cast<Exception>());
    }
  }
}
